crypto/bn/asm/rsaz-2k-avxifma.pl

```diff
@@ -0,0 +1,856 @@
+    vpsubq  .Lmask52x4(%rip), $_R0,  $T0
+    shl \$5, %r14
+    vmovapd (%rdx, %r14), $T1
+    vblendvpd $T1, $T0, $_R0, $_R0
+
+    shr \$4, %r13b
+    and \$0xf, %r13
+    vpsubq  .Lmask52x4(%rip), $_R0h,  $T0
+    shl \$5, %r13
+    vmovapd (%rdx, %r13), $T1
+    vblendvpd $T1, $T0, $_R0h, $_R0h
+
+    mov %r12b, %r11b
+    and \$0xf, %r12
+    vpsubq  .Lmask52x4(%rip), $_R1,  $T0
+    shl \$5, %r12
+    vmovapd (%rdx, %r12), $T1
+    vblendvpd $T1, $T0, $_R1, $_R1
+
+    shr \$4, %r11b
+    and \$0xf, %r11
+    vpsubq  .Lmask52x4(%rip), $_R1h,  $T0
+    shl \$5, %r11
+    vmovapd (%rdx, %r11), $T1
+    vblendvpd $T1, $T0, $_R1h, $_R1h
+
+    and \$0xf, %r10
+    vpsubq  .Lmask52x4(%rip), $_R2,  $T0
+    shl \$5, %r10
+    vmovapd (%rdx, %r10), $T1
+    vblendvpd $T1, $T0, $_R2, $_R2
+
+    # Add carries according to the obtained mask
+    vpand   .Lmask52x4(%rip), $_R0,  $_R0
+    vpand   .Lmask52x4(%rip), $_R0h, $_R0h
+    vpand   .Lmask52x4(%rip), $_R1,  $_R1
+    vpand   .Lmask52x4(%rip), $_R1h, $_R1h
+    vpand   .Lmask52x4(%rip), $_R2,  $_R2
+___
+}
+
+$code.=<<___;
+.text
+
+.globl  ossl_rsaz_amm52x20_x1_avxifma256
+.type   ossl_rsaz_amm52x20_x1_avxifma256,\@function,5
+.align 32
+ossl_rsaz_amm52x20_x1_avxifma256:
+.cfi_startproc
+    endbranch
+    push    %rbx
+.cfi_push   %rbx
+    push    %rbp
+.cfi_push   %rbp
+    push    %r12
+.cfi_push   %r12
+    push    %r13
+.cfi_push   %r13
+    push    %r14
+.cfi_push   %r14
+    push    %r15
+.cfi_push   %r15
+.Lossl_rsaz_amm52x20_x1_ifma256_body:
+
+    # Zeroing accumulators
+    vpxor    $zero, $zero, $zero
+    vmovapd   $zero, $R0_0
+    vmovapd   $zero, $R0_0h
+    vmovapd   $zero, $R1_0
+    vmovapd   $zero, $R1_0h
+    vmovapd   $zero, $R2_0
+
+    xorl    $acc0_0_low, $acc0_0_low
+
+    movq    $b, $b_ptr                       # backup address of b
+    movq    \$0xfffffffffffff, $mask52       # 52-bit mask
+
+    # Loop over 20 digits unrolled by 4
+    mov     \$5, $iter
+
+.align 32
+.Lloop5:
+___
+    foreach my $idx (0..3) {
+        &amm52x20_x1(0,8*$idx,$acc0_0,$R0_0,$R0_0h,$R1_0,$R1_0h,$R2_0,$k0);
+    }
+$code.=<<___;
+    lea    `4*8`($b_ptr), $b_ptr
+    dec    $iter
+    jne    .Lloop5
+___
+    &amm52x20_x1_norm($acc0_0,$R0_0,$R0_0h,$R1_0,$R1_0h,$R2_0);
+$code.=<<___;
+
+    vmovdqu   $R0_0,  `0*32`($res)
+    vmovdqu   $R0_0h, `1*32`($res)
+    vmovdqu   $R1_0,  `2*32`($res)
+    vmovdqu   $R1_0h, `3*32`($res)
+    vmovdqu   $R2_0,  `4*32`($res)
+
+    vzeroupper
+    mov  0(%rsp),%r15
+.cfi_restore    %r15
+    mov  8(%rsp),%r14
+.cfi_restore    %r14
+    mov  16(%rsp),%r13
+.cfi_restore    %r13
+    mov  24(%rsp),%r12
+.cfi_restore    %r12
+    mov  32(%rsp),%rbp
+.cfi_restore    %rbp
+    mov  40(%rsp),%rbx
+.cfi_restore    %rbx
+    lea  48(%rsp),%rsp
+.cfi_adjust_cfa_offset  -48
+.Lossl_rsaz_amm52x20_x1_ifma256_epilogue:
+    ret
+.cfi_endproc
+.size   ossl_rsaz_amm52x20_x1_avxifma256, .-ossl_rsaz_amm52x20_x1_avxifma256
+___
+
+$code.=<<___;
+.data
```

Comment:
I think this needs `.section .rodata align=64` to keep it working on OpenBSD.

Reply:
sorry I meant to say `align=32` as we need to change the section to .rodata. alignment should be kept. this is also for all other places, where I mistyped 64 instead of 32.